pkg/sources/github/repo.go

```diff
@@ -0,0 +1,232 @@
+		if err != nil {
+			return "", nil, fmt.Errorf("error cloning repo %s: %w", repoURL, err)
+		}
+
+	case *sourcespb.GitHub_GithubApp:
+		s.githubUser, s.githubToken, err = s.userAndToken(ctx, installationClient)
+		if err != nil {
+			return "", nil, fmt.Errorf("error getting token for repo %s: %w", repoURL, err)
+		}
+
+		path, repo, err = git.CloneRepoUsingToken(ctx, s.githubToken, repoURL, s.githubUser)
+		if err != nil {
+			return "", nil, fmt.Errorf("error cloning repo %s: %w", repoURL, err)
+		}
+
+	case *sourcespb.GitHub_Token:
+		// We never refresh user provided tokens, so if we already have them, we never need to try and fetch them again.
+		if s.githubUser == "" || s.githubToken == "" {
+			s.githubUser, s.githubToken, err = s.userAndToken(ctx, installationClient)
+			if err != nil {
+				return "", nil, fmt.Errorf("error getting token for repo %s: %w", repoURL, err)
+			}
+		}
+		path, repo, err = git.CloneRepoUsingToken(ctx, s.githubToken, repoURL, s.githubUser)
+		if err != nil {
+			return "", nil, fmt.Errorf("error cloning repo %s: %w", repoURL, err)
+		}
+	default:
+		return "", nil, fmt.Errorf("unhandled credential type for repo %s", repoURL)
+	}
+	return path, repo, nil
+}
+
+func (s *Source) userAndToken(ctx context.Context, installationClient *github.Client) (string, string, error) {
+	switch cred := s.conn.GetCredential().(type) {
+	case *sourcespb.GitHub_Unauthenticated:
+		// do nothing
+	case *sourcespb.GitHub_GithubApp:
+		id, err := strconv.ParseInt(cred.GithubApp.InstallationId, 10, 64)
+		if err != nil {
+			return "", "", fmt.Errorf("unable to parse installation id: %w", err)
+		}
+		// TODO: Check rate limit for this call.
+		token, _, err := installationClient.Apps.CreateInstallationToken(
+			ctx, id, &github.InstallationTokenOptions{})
+		if err != nil {
+			return "", "", fmt.Errorf("unable to create installation token: %w", err)
+		}
+		return "x-access-token", token.GetToken(), nil // TODO: multiple workers request this, track the TTL
+	case *sourcespb.GitHub_Token:
+		var (
+			ghUser *github.User
+			resp   *github.Response
+			err    error
+		)
+		for {
+			ghUser, resp, err = s.apiClient.Users.Get(context.TODO(), "")
+			if handled := s.handleRateLimit(err, resp); handled {
+				continue
+			}
+			if err != nil {
+				return "", "", fmt.Errorf("unable to get user: %w", err)
+			}
+			break
+		}
+		return ghUser.GetLogin(), cred.Token, nil
+	default:
+		return "", "", fmt.Errorf("unhandled credential type")
+	}
+
+	return "", "", fmt.Errorf("unhandled credential type")
+}
+
+type repoListOptions interface {
+	getListOptions() *github.ListOptions
+}
+
+type repoLister func(ctx context.Context, target string, opts repoListOptions) ([]*github.Repository, *github.Response, error)
+
+type appListOptions struct {
+	github.ListOptions
+}
+
+func (a *appListOptions) getListOptions() *github.ListOptions {
+	return &a.ListOptions
+}
+
+func (s *Source) addReposByApp(ctx context.Context) error {
```

Comment:
this method name starts with `add` but the other two very similar ones start with `get`. is there a reason for this? if not, i find the discrepancy pretty confusing.

Reply:
I think this was the original name, but you are correct, this should bet `get`. Thanks.